### DISCUSSION TITLE
Configure Mason with servers installed per project

### DISCUSSION THREAD

**Discussion by caneta** [mason-org/mason.nvim#1706] [Q&A]
Hi everyone, is it possible to configure Mason with servers installed per project?

I've read that [Mason installs its servers globally](https://github.com/williamboman/mason.nvim?tab=readme-ov-file#introduction), but I'm used to work with tsserver/eslint/prettier installed per project, because:

1. every project could have a different server version
2. every project could have a different server configuration
3. the configuration is shared into the project repo to assure that every developer involved uses the same version/conf

So, is there any alternative conf to achieve this?

Thank you

**Comment by powerman** [comment]
I'm planning to implement this setup by using autocmd triggered on events which changes current buffer/file (probably something like BufReadPre, BufNewFile, BufEnter, etc.) which will:
1. Remove from PATH previously prepended directory, if any.
2. Try to detect project root dir, do nothing if fail.
3. If exist project's bin dir inside a root dir then prepend project's bin/ dir to PATH and remember it for 1.

This should make project-specific tool versions to have priority over versions installed by mason.

**Reply by powerman** [reply]
After some more thinking on this it became clear: full-featured solution will be overly complicated.

- LSP design expects to have same LSP server process handle different workspaces (projects) instead of running different versions of same LSP server to handle files in different projects.
- Some LSP are just wrappers for other tools, but these LSP are still started just once, so we've just one attempt to provide them with `$PATH` specific to some project. To make them handle files from other project these LSP should implement such a feature (like `rust-analyzer` has implemented `checkOnSave.overrideCommand` option, as shown at https://github.com/neovim/nvim-lspconfig/wiki/Project-local-settings#configure-in-your-personal-settings-initlua). AFAIK most other LSP (e.g. `golangci_lint_ls` or `efm`) has no such a feature. And support switching between current project's files and other files unrelated to any project (e.g. in /tmp).

So, full-featured solution will have to auto-restart (some) LSP after changing `$PATH`. To make this work fast it'll have to also avoid restarting LSP when switching between files in same project and probably (in case project does not have different version of an LSP server itself in project-specific directory) avoid restarting LSP which are not depend on `$PATH`.

At the moment I decide to use trivial solution, which will detect project root and set `$PATH` just once, when Neovim starts (based on first opened file). Chances are it'll be enough for my workflow:
- I work on one project at a time;
- If I open files from another project I do this to just review something and not going to change these files (thus I don't need neither Formatters nor Linters/LSP for such files from other project at this point of time);
- If I switch to work on another project I restart Neovim in that project's directory.

Current implementation is in this [plugin](https://github.com/powerman/config.nvim/blob/4da21b5edd711b6a6b88cdc47cea68c3a965a170/lua/project/init.lua) added to Neovim config using this [commit](https://github.com/powerman/config.nvim/commit/4da21b5edd711b6a6b88cdc47cea68c3a965a170).

**Reply by caneta** [reply]
> At the moment I decide to use trivial solution, which will detect project root and set `$PATH` just once, when Neovim starts (based on first opened file). Chances are it'll be enough for my workflow:
> 
>     * I work on one project at a time;
> 
>     * If I open files from another project I do this to just review something and not going to change these files (thus I don't need neither Formatters nor Linters/LSP for such files from other project at this point of time);

This is not trivial at all and fits exactly my needs too. I was searching nothing more complicated than that; I'll give it a try, thanks!
